### PR TITLE
PLT-5597 add `marlowe-cardano-minimal` to enable benchmarking of Marlowe scripts

### PR DESCRIPTION
**Pull request by thealmarty**
This will be refactored in a follow up PR with some benchmarking components moved to `plutus-benchmark`.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Changelog fragments have been written (if appropriate)
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting master unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by michaelpj** [comment]
I'll take more of a look once you've got it working. I'd be tempted to add the benchmark itself and associated logic to `plutus-benchmark`. We have a bunch of budget tests = deterministic benchmarks in there, so I don't think it would be out of place, and it _is_ functioning as a benchmark for us.

**Comment by angerman** [comment]
force-pushed to re-excite ci.

File: marlowe-cardano-minimal/changelog.d/20230511_083148_marty.stumpf_plt_5597_benchmark_marlowe.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+### Added
+
+- Add the `marlowe-cardano-minimal` package for benchmarking.
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply marlowe-cardano-minimal/changelog.d/20230511_083148_marty.stumpf_plt_5597_benchmark_marlowe.md:3]: It's not published, we are never going to release versions of it?

File: marlowe-cardano-minimal/changelog.d/scriv.ini [hunk 2]
```diff
@@ -0,0 +1,3 @@
+[scriv]
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply marlowe-cardano-minimal/changelog.d/scriv.ini:1]: Yes, looks like a few of them aren't links.

**Comment by thealmarty** [comment]
Unfortunately some commits on master broke things again. Working on it.

File: marlowe-cardano-minimal/changelog.d/20230511_083148_marty.stumpf_plt_5597_benchmark_marlowe.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (thealmarty) [reply marlowe-cardano-minimal/changelog.d/20230511_083148_marty.stumpf_plt_5597_benchmark_marlowe.md:3]: If we aren't going to publish this then I'll remove the change log dir then.

File: marlowe-cardano-minimal/app/Benchmark/Marlowe/Types.hs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+
+-- | A benchmarking case.
+data Benchmark =
+  Benchmark
+  {
+    bDatum         :: Data  -- ^ The datum.
+  , bRedeemer      :: Data  -- ^ The redeemer.
+  , bScriptContext :: ScriptContext  -- ^ The script context.
+  , bReferenceCost :: Maybe ExBudget  -- ^ The previously measured execution costs.
```
Comment (michaelpj) [inline marlowe-cardano-minimal/app/Benchmark/Marlowe/Types.hs:33]: what specifically does this mean? the ones they measured on mainnet?

File: marlowe-cardano-minimal/app/Benchmark/Marlowe/RolePayout.hs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+-- | The hash for the Marlowe role-payout validator.
+validatorHash :: ScriptHash
+validatorHash = rolePayoutValidatorHash
+
+
+-- | The benchmark cases for the Marlowe role-payout validator.
+benchmarks :: IO (Either String [Benchmark])
+benchmarks =
+  let
+    update benchmark@Benchmark{..} =
```
Comment (michaelpj) [inline marlowe-cardano-minimal/app/Benchmark/Marlowe/RolePayout.hs:58]: what's going on here?

File: marlowe-cardano-minimal/app/Benchmark/Marlowe/RolePayout.hs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+          updateScriptHash
+            "e165610232235bbbbeff5b998b233daae42979dec92a6722d9cda989"
+            rolePayoutValidatorHash
+            bScriptContext
+      }
+  in
+    second (update <$>) <$> readBenchmarks "rolepayout"
+
+
+-- | An example benchmark for the Marlowe role-payout validator.
```
Comment (michaelpj) [inline marlowe-cardano-minimal/app/Benchmark/Marlowe/RolePayout.hs:70]: what's this for? debugging?

File: marlowe-cardano-minimal/app/Benchmark/Marlowe/Semantics.hs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+validatorHash :: ScriptHash
+validatorHash = marloweValidatorHash
+
+
+-- | The benchmark cases for the Marlowe semantics validator.
+benchmarks :: IO (Either String [Benchmark])
+benchmarks =
+  let
+    update benchmark@Benchmark{..} =
+      benchmark {
```
Comment (michaelpj) [inline marlowe-cardano-minimal/app/Benchmark/Marlowe/Semantics.hs:64]: ditto

File: marlowe-cardano-minimal/app/Benchmark/Marlowe/Semantics.hs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+            "e165610232235bbbbeff5b998b233daae42979dec92a6722d9cda989"
+            rolePayoutValidatorHash
+            bScriptContext
+      }
+  in
+    second (update <$>) <$> readBenchmarks "semantics"
+
+
+-- | An example benchmark for the Marlowe semantics validator.
+exampleBenchmark :: Benchmark
```
Comment (michaelpj) [inline marlowe-cardano-minimal/app/Benchmark/Marlowe/Semantics.hs:79]: ditto

File: marlowe-cardano-minimal/app/Benchmark/Marlowe/RolePayout.hs (hunk 5 above)
Comment (michaelpj) [reply marlowe-cardano-minimal/app/Benchmark/Marlowe/RolePayout.hs:70]: Generally it seems like we have a bunch of stuff that just supports the creation of these, whereas the "real" benchmarks we're just reading them from files?

**Review by michaelpj — APPROVED** [review]

**Comment by zliu41** [comment]
@bwbush I'm slightly confused. The README says

> Benchmarking: compare the budget before and after optimizations that the Plutus team implemented. It could be helpful to do the benchmarking as we implement the optimization even.

If our goal is to compare the budget, why are we checking in a bunch of flat files? Should't we be checking in the Plutus Tx code so that we can compare the budget before and after?

The flat files could be useful for benchmarking the performance of the UPLC evaluator, which is independent of the budget. Is that the reason for those flat files?

**Comment by zliu41** [comment]
> We are: that's the stuff in the src folder.

Right, so the purpose of the flat files is not clear.

File: marlowe-cardano-minimal/README.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+Marlowe is a platform for financial products as smart contracts. [Marlowe-Cardano](https://github.com/input-output-hk/marlowe-cardano) is an implementation of Marlowe for the Cardano blockchain, built on top of Plutus.
+
+In short, users write a Marlowe application in `Marlowe-Cardano`, which generates the corresponding code that is ready for the Plutus compiler with some manual revision. The code then goes through the Plutus compiler pipeline and gets executed on Cardano.
+
+The Plutus Core team has been working on optimizing the compiler such that a script's budget is reduced. The budget is a reflection of
+
+(1) script sizes
+(2) execution costs
+
+It would be informative for both the Plutus and Marlowe teams to investigate in detail how the Marlowe scripts can be optimized. In particular, we can perform:
```
[the remark this answers is not in the document]
Comment (thealmarty) [reply marlowe-cardano-minimal/README.md:14]: This project is to enable those plans? I thought it makes sense to mention why we have this package here.

File: marlowe-cardano-minimal/README.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+The following files are output:
+- For Marlowe's semantics valdator
+    - Plutus script: `marlowe-semantics.plutus`
+    - Benchmarking results: `marlowe-semantics.tsv`   
+- For Marlowe's role-payout valdator
+    - Plutus script: `marlowe-rolepayout.plutus`
+    - Benchmarking results: `marlowe-rolepayout.tsv`   
+
+
+See [analysis.ipynb](analysis.ipynb) for example plots of the results.
```
[the remark this answers is not in the document]
Comment (thealmarty) [reply marlowe-cardano-minimal/README.md:71]: Reworded. Frankly I'll have a follow up PR moving some of the benchmarking stuff to `plutus-benchmark`. I'm not sure if we still want to keep the executable. Could be useful.

**Comment by thealmarty** [comment]
The same error came back after I rebase on master again :(
```
Resolving dependencies...
Error: cabal: Could not resolve dependencies:
[__0] trying: marlowe-cardano-minimal-0.1.0.3 (user goal)
[__1] next goal: plutus-core (user goal)
[__1] rejecting: plutus-core-1.7.0.0 (conflict: marlowe-cardano-minimal =>
plutus-core^>=1.6)
[__1] skipping: plutus-core-1.5.0.1, plutus-core-1.5.0.0, plutus-core-1.4.0.0,
plutus-core-1.3.0.0, plutus-core-1.2.0.0, plutus-core-1.1.1.0,
plutus-core-1.1.0.0, plutus-core-1.0.0.1, plutus-core-1.0.0.0 (has the same
characteristics that caused the previous version to fail: excluded by
constraint '^>=1.6' from 'marlowe-cardano-minimal')
[__1] fail (backjumping, conflict set: marlowe-cardano-minimal, plutus-core)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: marlowe-cardano-minimal, plutus-core
```
It was fixed last time after rebasing on #5333 - investigating more.

**Comment by thealmarty** [comment]
OK nvm. Updated cabal file to 1.7.0.0

**Pull request merged**